docs/options.md
# Widget Options

## Table of contents
- [Default Options](https://github.com/ammein/wagtail-custom-code-editor/blob/main/docs/options.md#default-options)
- [Copy & Paste Options](https://github.com/ammein/wagtail-custom-code-editor/blob/main/docs/options.md#copy--paste-options)
- [Reset Options](https://github.com/ammein/wagtail-custom-code-editor/blob/main/docs/options.md#reset-options)

### Default Options

```python
mode="html",
theme="chrome",
width="100%",
height="500px",
font_size=None,
keybinding=None,
useworker=True,
extensions=None,
enable_options=True,
enable_modes=False,
dropdown_config=None,
read_only_config=None,
save_command_config=None,
options=None,
modes=None,
attrs=None
```

- `mode` : `<str>`
> Set default mode when code editor appear. Default is `html`.
- `theme` : `<str>` 
> Set default theme. Default is `chrome`. See [list of themes](https://github.com/ajaxorg/ace-builds/blob/master/src/ext-themelist.js#L9).
- `width` : `<str>` 
> Width of Ace Editor. Default is `100%`.
- `height` : `<str>` 
> Height of Ace Editor. Default is `500px`.
- `font_size` : `<str or None>`
> Font size of ace editor
- `keybinding` : `<str or None>`
> Type of keybinding option for Ace Editor. You must insert the value without `keybinding-` and `.js` into `keybinding` value. Ex: `keybinding="emacs"`.
- `useworker` : `<bool>` 
> Turn On/Off for Ace Editor either to use Worker or not (Worker only works on certain mode. [Check here to see which mode supports worker](https://github.com/ajaxorg/ace-builds/tree/master/src)). Default is `True`.
- `extensions` : `<List[str] or str or None>` 
> Ace Editor extensions to use (All extensions is used by default). You must insert the value without `ext-` and `.js` . Ex: `extensions=["hardwrap", "keybinding_menu", "linking"]`. Check [list of extensions](https://github.com/ajaxorg/ace/tree/master/src/ext).
- `enable_options` : `<bool>` 
> Turn On/Off for Ace Editor Options Configurations that appear as a drawer on the right container. Default is `True`.
- `enable_modes` : `<bool>` 
> Turn On/Off for Ace Editor change mode feature when clicking the dropdown. Default is `False`.
- `dropdown_config` : `<Dict[str, str] or None>`
> CSS Config for Dropdown.
```python
dropdown_config={
    "position": {
        "top": "<string>",
        "left": "<string>",
        "right": "<string>",
        "bottom": "<string>"
    },
    "height": "<string>",
    "borderRadius": "<string>",
    "boxShadow": "<string>",
    "width": "<string>",
    "backgroundColor": "<string>"
}
```
- `read_only_config` : `<Dict[str, str] or None>`
> CSS Config for Read Only container.
```python
read_only_config={
    "position": {
        "top": "<string>",
        "left": "<string>",
        "right": "<string>",
        "bottom": "<string>"
    },
    "height": "<string>",
    "borderRadius": "<string>",
    "boxShadow": "<string>",
    "width": "<string>",
    "backgroundColor": "<string>"
}
```
- `save_command_config` : `<Dict[str, str] or None>`
> Save Command Keys for mac & windows when use switch mode. All ace editor [keyboard shortcuts here](https://github.com/ajaxorg/ace/wiki/Default-Keyboard-Shortcuts)
```python
# Default Value
save_command_config={
    "win": "Alt-Shift-S",
    "mac": "Option-Shift-S",
}
```
- `options` : `<Dict[str, Any] or str or None>`
> Override existing Ace Editor Options using key and value. Ex: `options=dict(readOnly=True, cursorStyle="smooth", printMarginColumn=100, showFoldWidgets=True, printMargin=50)`
- `modes` : `<List[Dict[str, str]] or None>`
> Override/extend Ace Editor Modes for switching modes feature. Refer [mode's properties](https://github.com/ammein/wagtail-custom-code-editor/blob/main/docs/settings.md#modes-properties). Ex: `modes=list({ "title": "GLSL", "name": "glsl", "snippet": "@code-here" })`. Check [list of modes](https://github.com/ajaxorg/ace-builds/blob/master/src/ext-modelist.js#L36).
- `attrs` : `<Dict[str, Any] or None>`
> You can set any html value here for the editor container. https://docs.djangoproject.com/en/5.1/ref/forms/widgets/#django.forms.Widget.attrs

### Copy & Paste Options
The best thing about this custom code editor is that you can test out your options in your code editor, copy it and paste the options in your widget/block.

![Copy & Paste Options](https://raw.githubusercontent.com/ammein/wagtail-custom-code-editor/refs/heads/main/docs/options.gif)

Then you can paste the copied options into the `options` kwargs like this:
```python
from wagtail_custom_code_editor.blocks import CustomCodeEditorBlock
from wagtail.blocks import (
    StructBlock,
)

class CodeBlock(StructBlock):
    code = CustomCodeEditorBlock(
                mode='c_cpp', 
                theme='chrome', 
                options="{\"highlightActiveLine\":false,\"showLineNumbers\":false}"
            )
```

> You can also use `Dict` in `options` kwargs like this: `options=dict(highlightActiveLine=False, showLineNumbers=False)`

### Reset Options
Reset options is as easy as press the second button near to copy button like this:

![Reset Options](https://raw.githubusercontent.com/ammein/wagtail-custom-code-editor/refs/heads/main/docs/reset.gif)

docs/settings.md
# Settings

## Table of Contents
- [Override Modes](https://github.com/ammein/wagtail-custom-code-editor/blob/main/docs/settings.md#override-modes)
  - [Modes Properties](https://github.com/ammein/wagtail-custom-code-editor/blob/main/docs/settings.md#modes-properties)
- [Emmet URL](https://github.com/ammein/wagtail-custom-code-editor/blob/main/docs/settings.md#emmet-url)
- [Ace Editor Options Settings](https://github.com/ammein/wagtail-custom-code-editor/blob/main/docs/settings.md#ace-editor-options-settings)
- [Default Settings](https://github.com/ammein/wagtail-custom-code-editor/blob/main/docs/settings.md#default-settings)

### Override Modes
These modes are used when you set kwargs parameter in widget called `enable_modes=True`, and you can easily switch mode by clicking the dropdown button. You can override the whole modes using _List_ of _Dictionaries_ from key named `MODES` in settings:
```python
WAGTAIL_CUSTOM_CODE_EDITOR = {
    "MODES": [
        {
            "title": "Installed Apps Django",
            "name": "python",
            "snippet": """
            INSTALLED_APPS = [
                @code-here
            ]
            """
        },
        {
            "title": "XML",
            "name": "xml",
            "snippet": """
            <note>
                <script/>
                <to>Tove</to>
                <from>Jani</from>
                <heading>Reminder</heading>
                <body>@code-here</body>
            </note>
            """
        }
    ]
}
```
#### Modes Properties
- `title` - You can set any title name you want for specific mode
- `name` - (CASE SENSITIVE) You must set the name of the mode for Ace Editor to set appropriate mode. You can see list of latest [modes here](https://github.com/ajaxorg/ace/tree/master/src/mode).
> You must insert the value for `name` without `mode-` and `.js`. 
- `snippet` - This is where your existing code will be replaced by this snippet value.
- `disableSnippet` - You can set this key using boolean value if you want to disable the snippet on specific mode's name.
- `@code-here` - If you notice this from example above, any existing code will be placed on `@code-here` placement anywhere in the snippet texts. Useful if you wish retain the same value when switching modes.

### Emmet URL
By default, the emmet used for `html` mode is for experienced programmers who build HTML faster using emmet shortcodes. This needs to be extended for Ace Editor usage, and you can provide any emmet extension here by replacing this value:
```python
WAGTAIL_CUSTOM_CODE_EDITOR = {
    "EMMET_URL": "https://custom-emmet-extensions/emmet.js"
}
```

If you wish to push your own emmet extensions or disable emmet extensions, you can simply set to `None` like this:
```python
WAGTAIL_CUSTOM_CODE_EDITOR = {
    "EMMET_URL": None
}
```

### Ace Editor Options Settings
These options settings are the inputs for you to configure ace options from the Code Editor Field in your admin. You will be able to see these options when you clicked _Cog_ icon on top right corner (Available when `enable_options` is set to `True`). If you wish to override options settings, you may configure it in settings like this:
```python
WAGTAIL_CUSTOM_CODE_EDITOR = {
    "OPTIONS_TYPES": [
        # For Select Input
        {
            "name": "<your-custom-ace-options-name>", # Ace options name
            "type": "string", # Type of input value
            "value": ["line", "text"], # Choices value
            "category": "editor", # Category for this options
        },
        # For Select Custom Title Input
        {
            "name": "<your-custom-ace-options-name>",
            "type": "string", # Type of input value
            "value": [
                {"title": "Always", "value": "always"},
                {"title": "Never", "value": False},
                {"title": "Timed", "value": True},
            ], # Choices value
            "category": "editor", # Category for this options
        },
        # For Number Input
        {
            "name": "<your-custom-ace-options-name>", # Ace options name
            "type": "number", # Type of input value
            "value": 10, # Default Value
            "category": "mouseHandler", # Category for this options
        },
        # For Slider Number Input
        {
            "name": "<your-custom-ace-options-name>",
            "type": "number",  # Type of input value
            "value": {"min": 0, "max": 100, "steps": 1}, # Default Slider Attribute Value
            "category": "renderer", # Category for this options
        },
    ]
}
```

### 

### Default Settings

```python
WAGTAIL_CUSTOM_CODE_EDITOR = {
    "EMMET_URL": "https://cloud9ide.github.io/emmet-core/emmet.js",
    # MODES
    "MODES": [
        {
            "title": "Bash",
            "name": "sh",
            "snippet": """
                #!/bin/bash
                # GNU bash, version 4.3.46
                @code-here
            """
        },
        {
            "title": "ActionScript",
            "name": "actionscript"
        },
        {
            "title": "C++",
            "name": "c_cpp",
            "snippet": """
                //Microsoft (R) C/C++ Optimizing Compiler Version 19.00.23506 for x64
                #include <iostream>
        
                int main()
                {
                   @code-here
                }
            """
        },
        {
            "title": "C#",
            "name": "csharp",
            "snippet": """
                //Rextester.Program.Main is the entry point for your code. Don't change it.
                //Compiler version 4.0.30319.17929 for Microsoft (R) .NET Framework 4.5
            
                using System;
                using System.Collections.Generic;
                using System.Linq;
                using System.Text.RegularExpressions;
            
                namespace Rextester
                {
                   public class Program
                   {
                      public static void Main(string[] args)
                      {
                            // code goes here
                            @code-here
                      }
                   }
                }
            """
        },
        {
            "name": "php",
            "snippet": """
                <html>
                <head>
                <title>PHP Test</title>
                </head>
                <body>
                <?php //code goes here
                   @code-here
                ?> 
                </body>
                </html>
            """
        },
        {
            "name": "html",
            "snippet": """
                <!DOCTYPE html>
                <html>
                <head>
                <title>
                    <!-- Title Here -->
                </title>
                </head>
                <body>
                    <!-- Code-here -->
                    @code-here
                </body>
                </html>
            """
        },
        {
            "name": "javascript",
            "snippet": """
                document.addEventListener("DOMContentLoaded" , function(){
                   @code-here
                });
            """
        }
    ],
    # Options
    "OPTIONS_TYPES": [
        {
            "name": "selectionStyle",
            "type": "string",
            "value": ["line", "text"],
            "category": "editor",
        },
        {
            "name": "highlightActiveLine",
            "type": "boolean",
            "value": None,
            "category": "editor",
        },
        {
            "name": "highlightSelectedWord",
            "type": "boolean",
            "value": None,
            "category": "editor",
        },
        {
            "name": "readOnly",
            "type": "boolean",
            "value": None,
            "category": "editor",
        },
        {
            "name": "cursorStyle",
            "type": "string",
            "value": ["ace", "slim", "smooth", "wide"],
            "category": "editor",
        },
        {
            "name": "mergeUndoDeltas",
            "type": "string",
            "value": [
                {"title": "Always", "value": "always"},
                {"title": "Never", "value": False},
                {"title": "Timed", "value": True},
            ],
            "category": "editor",
        },
        {
            "name": "behavioursEnabled",
            "type": "boolean",
            "value": None,
            "category": "editor",
        },
        {
            "name": "wrapBehavioursEnabled",
            "type": "boolean",
            "value": None,
            "category": "editor",
        },
        {
            "name": "autoScrollEditorIntoView",
            "type": "boolean",
            "value": None,
            "help": "This is needed if editor is inside scrollable page",
            "category": "editor",
        },
        {
            "name": "copyWithEmptySelection",
            "type": "boolean",
            "value": None,
            "help": "Copy/Cut the full line if selection is empty, "
                    "defaults to False",
            "category": "editor",
        },
        {
            "name": "useSoftTabs",
            "type": "boolean",
            "value": None,
            "category": "editor",
        },
        {
            "name": "navigateWithinSoftTabs",
            "type": "boolean",
            "value": None,
            "category": "editor",
        },
        {
            "name": "enableMultiSelect",
            "type": "boolean",
            "value": None,
            "category": "editor",
        },
        {
            "name": "hScrollBarAlwaysVisible",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "vScrollBarAlwaysVisible",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "highlightGutterLine",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "animatedScroll",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "showInvisibles",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "showPrintMargin",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "printMarginColumn",
            "type": "number",
            "value": {"min": 0, "max": 100, "steps": 1},
            "category": "renderer",
        },
        {
            "name": "printMargin",
            "type": "number",
            "value": {"min": 0, "max": 100, "steps": 1},
            "category": "renderer",
        },
        {
            "name": "fadeFoldWidgets",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "showFoldWidgets",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "showLineNumbers",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "showGutter",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "displayIndentGuides",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "scrollPastEnd",
            "type": ["number", "boolean"],
            "value": {"min": 0, "max": 1, "steps": 0.1},
            "help": "Number of page sizes to scroll after document end "
                    "(typical values are 0, 0.5, and 1)",
            "category": "renderer",
        },
        {
            "name": "fixedWidthGutter",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "scrollSpeed",
            "type": "number",
            "value": {"min": 0, "max": 100, "steps": 1},
            "category": "mouseHandler",
        },
        {
            "name": "dragDelay",
            "type": "number",
            "value": {"min": 0, "max": 200, "steps": 1},
            "category": "mouseHandler",
        },
        {
            "name": "dragEnabled",
            "type": "boolean",
            "value": None,
            "category": "mouseHandler",
        },
        {
            "name": "focusTimeout",
            "type": "number",
            "value": None,
            "category": "mouseHandler",
        },
        {
            "name": "tooltipFollowsMouse",
            "type": "boolean",
            "value": None,
            "category": "mouseHandler",
        },
        {
            "name": "overwrite",
            "type": "boolean",
            "value": None,
            "category": "session",
        },
        {
            "name": "newLineMode",
            "type": "string",
            "value": ["auto", "unix", "windows"],
            "category": "session",
        },
        {
            "name": "tabSize",
            "type": "number",
            "value": {"min": 0, "max": 20, "steps": 1},
            "category": "session",
        },
        {
            "name": "wrap",
            "type": ["boolean", "number"],
            "value": None,
            "category": "session",
        },
        {
            "name": "foldStyle",
            "type": "string",
            "value": ["markbegin", "markbeginend", "manual"],
            "category": "session",
        },
        {
            "name": "enableBasicAutocompletion",
            "type": "boolean",
            "value": None,
            "category": "extension",
        },
        {
            "name": "enableLiveAutocompletion",
            "type": "boolean",
            "value": None,
            "category": "extension",
        },
        {
            "name": "enableEmmet",
            "type": "boolean",
            "value": None,
            "category": "extension",
        },
        {
            "name": "useElasticTabstops",
            "type": "boolean",
            "value": None,
            "category": "extension",
        },
    ]
}
```

wagtail_custom_code_editor/settings.py
import os
import warnings
from django.conf import settings
from django.test.signals import setting_changed

DEFAULTS = {
    "EMMET_URL": "https://cloud9ide.github.io/emmet-core/emmet.js",
    # MODES
    "MODES": [
        {
            "title": "Bash",
            "name": "sh",
            "snippet": """
                #!/bin/bash
                # GNU bash, version 4.3.46
                @code-here
            """
        },
        {
            "title": "ActionScript",
            "name": "actionscript"
        },
        {
            "title": "C++",
            "name": "c_cpp",
            "snippet": """
                //Microsoft (R) C/C++ Optimizing Compiler Version 19.00.23506 for x64
                #include <iostream>
        
                int main()
                {
                   @code-here
                }
            """
        },
        {
            "title": "C#",
            "name": "csharp",
            "snippet": """
                //Rextester.Program.Main is the entry point for your code. Don't change it.
                //Compiler version 4.0.30319.17929 for Microsoft (R) .NET Framework 4.5
            
                using System;
                using System.Collections.Generic;
                using System.Linq;
                using System.Text.RegularExpressions;
            
                namespace Rextester
                {
                   public class Program
                   {
                      public static void Main(string[] args)
                      {
                            // code goes here
                            @code-here
                      }
                   }
                }
            """
        },
        {
            "name": "php",
            "snippet": """
                <html>
                <head>
                <title>PHP Test</title>
                </head>
                <body>
                <?php //code goes here
                   @code-here
                ?> 
                </body>
                </html>
            """
        },
        {
            "name": "html",
            "snippet": """
                <!DOCTYPE html>
                <html>
                <head>
                <title>
                    <!-- Title Here -->
                </title>
                </head>
                <body>
                    <!-- Code-here -->
                    @code-here
                </body>
                </html>
            """
        },
        {
            "name": "javascript",
            "snippet": """
                document.addEventListener("DOMContentLoaded" , function(){
                   @code-here
                });
            """
        }
    ],
    # Options
    "OPTIONS_TYPES": [
        {
            "name": "selectionStyle",
            "type": "string",
            "value": ["line", "text"],
            "category": "editor",
        },
        {
            "name": "highlightActiveLine",
            "type": "boolean",
            "value": None,
            "category": "editor",
        },
        {
            "name": "highlightSelectedWord",
            "type": "boolean",
            "value": None,
            "category": "editor",
        },
        {
            "name": "readOnly",
            "type": "boolean",
            "value": None,
            "category": "editor",
        },
        {
            "name": "cursorStyle",
            "type": "string",
            "value": ["ace", "slim", "smooth", "wide"],
            "category": "editor",
        },
        {
            "name": "mergeUndoDeltas",
            "type": "string",
            "value": [
                {"title": "Always", "value": "always"},
                {"title": "Never", "value": False},
                {"title": "Timed", "value": True},
            ],
            "category": "editor",
        },
        {
            "name": "behavioursEnabled",
            "type": "boolean",
            "value": None,
            "category": "editor",
        },
        {
            "name": "wrapBehavioursEnabled",
            "type": "boolean",
            "value": None,
            "category": "editor",
        },
        {
            "name": "autoScrollEditorIntoView",
            "type": "boolean",
            "value": None,
            "help": "This is needed if editor is inside scrollable page",
            "category": "editor",
        },
        {
            "name": "copyWithEmptySelection",
            "type": "boolean",
            "value": None,
            "help": "Copy/Cut the full line if selection is empty, "
                    "defaults to False",
            "category": "editor",
        },
        {
            "name": "useSoftTabs",
            "type": "boolean",
            "value": None,
            "category": "editor",
        },
        {
            "name": "navigateWithinSoftTabs",
            "type": "boolean",
            "value": None,
            "category": "editor",
        },
        {
            "name": "enableMultiSelect",
            "type": "boolean",
            "value": None,
            "category": "editor",
        },
        {
            "name": "hScrollBarAlwaysVisible",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "vScrollBarAlwaysVisible",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "highlightGutterLine",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "animatedScroll",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "showInvisibles",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "showPrintMargin",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "printMarginColumn",
            "type": "number",
            "value": {"min": 0, "max": 100, "steps": 1},
            "category": "renderer",
        },
        {
            "name": "printMargin",
            "type": "number",
            "value": {"min": 0, "max": 100, "steps": 1},
            "category": "renderer",
        },
        {
            "name": "fadeFoldWidgets",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "showFoldWidgets",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "showLineNumbers",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "showGutter",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "displayIndentGuides",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "scrollPastEnd",
            "type": ["number", "boolean"],
            "value": {"min": 0, "max": 1, "steps": 0.1},
            "help": "Number of page sizes to scroll after document end "
                    "(typical values are 0, 0.5, and 1)",
            "category": "renderer",
        },
        {
            "name": "fixedWidthGutter",
            "type": "boolean",
            "value": None,
            "category": "renderer",
        },
        {
            "name": "scrollSpeed",
            "type": "number",
            "value": {"min": 0, "max": 100, "steps": 1},
            "category": "mouseHandler",
        },
        {
            "name": "dragDelay",
            "type": "number",
            "value": {"min": 0, "max": 200, "steps": 1},
            "category": "mouseHandler",
        },
        {
            "name": "dragEnabled",
            "type": "boolean",
            "value": None,
            "category": "mouseHandler",
        },
        {
            "name": "focusTimeout",
            "type": "number",
            "value": None,
            "category": "mouseHandler",
        },
        {
            "name": "tooltipFollowsMouse",
            "type": "boolean",
            "value": None,
            "category": "mouseHandler",
        },
        {
            "name": "overwrite",
            "type": "boolean",
            "value": None,
            "category": "session",
        },
        {
            "name": "newLineMode",
            "type": "string",
            "value": ["auto", "unix", "windows"],
            "category": "session",
        },
        {
            "name": "tabSize",
            "type": "number",
            "value": {"min": 0, "max": 20, "steps": 1},
            "category": "session",
        },
        {
            "name": "wrap",
            "type": ["boolean", "number"],
            "value": None,
            "category": "session",
        },
        {
            "name": "foldStyle",
            "type": "string",
            "value": ["markbegin", "markbeginend", "manual"],
            "category": "session",
        },
        {
            "name": "enableBasicAutocompletion",
            "type": "boolean",
            "value": None,
            "category": "extension",
        },
        {
            "name": "enableLiveAutocompletion",
            "type": "boolean",
            "value": None,
            "category": "extension",
        },
        {
            "name": "enableEmmet",
            "type": "boolean",
            "value": None,
            "category": "extension",
        },
        {
            "name": "useElasticTabstops",
            "type": "boolean",
            "value": None,
            "category": "extension",
        },
    ]
}

# List of settings that have been deprecated
DEPRECATED_SETTINGS = []

# List of settings that have been removed
# note: use a tuple of (setting, deprecation warning from deprecation.py)
REMOVED_SETTINGS = []


class CustomCodeEditorSettings:
    """
    A settings object that allows the wagtailmedia settings to be accessed as
    properties. For example:
        from wagtail_custom_code_editor.settings import customcodeeditor_settings
        print(wagtailmedia_settings.MEDIA_MODEL)
    Note:
    This is an internal class that is only compatible with settings namespaced
    under the WAGTAILCUSTOMCODEEDITOR name. It is not intended to be used by 3rd-party
    apps, and test helpers like `override_settings` may not work as expected.
    """

    def __init__(self, user_settings=None, defaults=None):
        if user_settings:
            self._user_settings = self.__check_user_settings(user_settings)
        self.defaults = defaults or DEFAULTS
        self._cached_attrs = set()

    @property
    def user_settings(self):
        if not hasattr(self, "_user_settings"):
            self._user_settings = self.__check_user_settings(
                getattr(settings, "WAGTAIL_CUSTOM_CODE_EDITOR", {})
            )
        return self._user_settings

    def __getattr__(self, item):
        if item not in self.defaults:
            raise AttributeError(f"Invalid wagtail_custom_code_editor setting: '{item}'")

        try:
            # Check if present in user settings
            val = self.user_settings[item]
        except KeyError:
            val = self.defaults[item]

        # Cache the result
        self._cached_attrs.add(item)
        setattr(self, item, val)
        return val

    @staticmethod
    def __check_user_settings(user_settings):
        for setting, category in DEPRECATED_SETTINGS:
            if setting in user_settings or hasattr(settings, setting):
                warnings.warn(
                    f"The '{setting}' setting is deprecated and will be removed in the next release, ",
                    category=category,
                    stacklevel=2,
                )

        for setting in REMOVED_SETTINGS:
            if setting in user_settings:
                raise RuntimeError(
                    f"The '{setting}' setting has been removed."
                    f"Please refer to the wagtail_custom_code_editor documentation for available settings."
                )

        return user_settings

    def reload(self):
        for attr in self._cached_attrs:
            delattr(self, attr)
        self._cached_attrs.clear()
        if hasattr(self, "_user_settings"):
            delattr(self, "_user_settings")


wagtail_custom_code_editor_settings = CustomCodeEditorSettings(None, DEFAULTS)



def reload_wagtail_custom_code_editor_settings(**kwargs):
    setting = kwargs['setting']
    if setting == "WAGTAIL_CUSTOM_CODE_EDITOR":
        wagtail_custom_code_editor_settings.reload()


setting_changed.connect(reload_wagtail_custom_code_editor_settings)
